File: 백준/Gold/11066. 파일 합치기/파일 합치기.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int T = Integer.parseInt(br.readLine());
        
        for (int i = 0; i < T; i++) {
            int K = Integer.parseInt(br.readLine());
            int[] sum = new int[K+1];
            int[][] dp = new int[K+1][K+1];
            st = new StringTokenizer(br.readLine(), " ");
            for (int j = 1; j <= K; j++) {
                sum[j] = sum[j-1] + Integer.parseInt(st.nextToken());
            }
            
            for (int a = 1; a < K; a++) {
                for (int start = 1; start + a <= K; start++) {
                    int end = start + a;
                    dp[start][end] = Integer.MAX_VALUE;
                    for (int mid = start; mid < end; mid++) {
                        dp[start][end] 
                            = Math.min(dp[start][mid] + dp[mid+1][end] + 
                                       sum[end] - sum[start-1], dp[start][end]);
                    }
                }
            }
            System.out.println(dp[1][K]);
        }
    }
}

File: 백준/Gold/11066. 파일 합치기/파일 합치기.cc
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

int main() {
    int T;
    cin >> T;

    for (int i = 0; i < T; i++) {
        int K;
        cin >> K;
        vector<int> files(K);
        vector<int> sum(K+1, 0);
        vector<vector<int>> dp(K+1, vector<int>(K+1, 0));

        for (int j = 0; j < K; j++) {
            cin >> files[j];
        }

        for (int j = 1; j <= K; j++) {
            sum[j] = sum[j-1] + files[j-1];
        }

        for (int a = 1; a < K; a++) {
            for (int start = 1; start + a <= K; start++) {
                int end = start + a;
                dp[start][end] = INT_MAX;
                for (int mid = start; mid < end; mid++) {
                    dp[start][end] 
                        = min(dp[start][end], dp[start][mid] + dp[mid+1][end] 
                              + sum[end] - sum[start-1]);
                }
            }
        }

        cout << dp[1][K] << endl;
    }

    return 0;
}